homeassistant/components/picnic/coordinator.py

```diff
@@ -0,0 +1,142 @@
+"""Coordinator to fetch data from the Picnic API."""
+import copy
+from datetime import timedelta
+import logging
+
+import async_timeout
+from python_picnic_api import PicnicAPI
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    ADDRESS,
+    SENSOR_CART_ITEMS_COUNT,
+    SENSOR_CART_TOTAL_PRICE,
+    SENSOR_COMPLETED_DELIVERIES,
+    SENSOR_LAST_ORDER_DELIVERY_TIME,
+    SENSOR_LAST_ORDER_ETA_END,
+    SENSOR_LAST_ORDER_ETA_START,
+    SENSOR_LAST_ORDER_SLOT_END,
+    SENSOR_LAST_ORDER_SLOT_START,
+    SENSOR_LAST_ORDER_STATUS,
+    SENSOR_LAST_ORDER_TOTAL_PRICE,
+    SENSOR_SELECTED_SLOT_END,
+    SENSOR_SELECTED_SLOT_MAX_ODER_TIME,
+    SENSOR_SELECTED_SLOT_MIN_ORDER_VALUE,
+    SENSOR_SELECTED_SLOT_START,
+    SENSOR_TOTAL_ORDERS,
+)
+
+
+class PicnicUpdateCoordinator(DataUpdateCoordinator):
+    """The coordinator to fetch data from the Picnic API at a set interval."""
+
+    def __init__(self, hass: HomeAssistant, picnic_api_client: PicnicAPI):
+        """Initialize the coordinator with the given Picnic API client."""
+        self.picnic_api_client = picnic_api_client
+
+        logger = logging.getLogger(__name__)
+        super().__init__(
+            hass,
+            logger,
+            name="Picnic coordinator",
+            update_interval=timedelta(minutes=30),
+            update_method=self.async_update_data,
+        )
+
+    async def async_update_data(self):
+        """Fetch data from API endpoint."""
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                return await self.hass.async_add_executor_job(self.fetch_data)
+        except Exception as err:  # pylint: disable=broad-except
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+    def fetch_data(self):
+        """Fetch the data from the Picnic API and return a flat dict with only needed sensor data."""
+        # Fetch from the API and pre-process the data
+        user = self.picnic_api_client.get_user()
+        cart = self.picnic_api_client.get_cart()
+        last_order = self._get_last_order()
+
+        if not user or not cart or not last_order:
+            raise UpdateFailed("API response doesn't contain expected data.")
+
+        slot_data = self._get_slot_data(cart)
+        address = f'{user["address"]["street"]} {user["address"]["house_number"]}{user["address"]["house_number_ext"]}'
+        minimum_order_value = (
+            slot_data["minimum_order_value"] / 100
+            if slot_data.get("minimum_order_value")
+            else None
+        )
+        # Create a flat lookup table to be used in the entities, convert prices from cents to euros
```

Comment:
Move this logic to `SENSOR_TYPES` as part of each sensor definition.

Reply:
I'm not totally sure what I should change here, do you mean moving the logic of creating the lookup table to the sensor definition or the conversion from cents to euros, or both? And wouldn't that require more memory so save the whole API responses instead of the data we're using?

Reply:
Instead of creating the lookup table, have each sensor definition have a lambda that knows how to extract the data it needs. Like this https://github.com/home-assistant/core/blob/dev/homeassistant/components/shelly/sensor.py#L34

Reply:
I've changed it according to your suggestion, but please check that it's done in the right way 😃.